CMakeLists.txt

```diff
@@ -636,6 +635,119 @@ if(CONFIG_GEN_ISR_TABLES)
   set_property(GLOBAL APPEND PROPERTY GENERATED_KERNEL_SOURCE_FILES isr_tables.c)
 endif()
 
+if(CONFIG_ARM_USERSPACE)
+  set(GEN_PRIV_STACKS $ENV{ZEPHYR_BASE}/scripts/gen_priv_stacks.py)
+  set(PROCESS_PRIV_STACKS_GPERF $ENV{ZEPHYR_BASE}/scripts/process_gperf.py)
+
+  set(PRIV_STACKS                    priv_stacks_hash.gperf)
+  set(PRIV_STACKS_OUTPUT_SRC_PRE     priv_stacks_hash_preprocessed.c)
+  set(PRIV_STACKS_OUTPUT_SRC         priv_stacks_hash.c)
+  set(PRIV_STACKS_OUTPUT_OBJ         priv_stacks_hash.c.obj)
+  set(PRIV_STACKS_OUTPUT_OBJ_RENAMED priv_stacks_hash_renamed.o)
+
+  # Essentially what we are doing here is extracting some information
+  # out of the nearly finished elf file, generating the source code
+  # for a hash table based on that information, and then compiling and
+  # linking the hash table back into a now even more nearly finished
+  # elf file.
+
+  # Use the script GEN_PRIV_STACKS to scan the kernel binary's
+  # (zephyr_prebuilt) DWARF information to produce a table of kernel
+  # objects (PRIV_STACKS) which we will then pass to gperf
+  add_custom_command(
+    OUTPUT ${PRIV_STACKS}
+    COMMAND
+    ${PYTHON_EXECUTABLE}
+    ${GEN_PRIV_STACKS}
+    --kernel $<TARGET_FILE:priv_stacks_prebuilt>
+    --output ${PRIV_STACKS}
+    $<$<BOOL:${CMAKE_VERBOSE_MAKEFILE}>:--verbose>
+    DEPENDS priv_stacks_prebuilt
+    WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
+    )
+  add_custom_target(priv_stacks DEPENDS ${CMAKE_CURRENT_BINARY_DIR}/${PRIV_STACKS})
+
+  # Use gperf to generate C code (PRIV_STACKS_OUTPUT_SRC_PRE) which implements a
+  # perfect hashtable based on PRIV_STACKS
+  add_custom_command(
+    OUTPUT ${PRIV_STACKS_OUTPUT_SRC_PRE}
+    COMMAND
+    ${GPERF}
```

Comment:
add -C so we can put this stuff completely in ROM

Reply:
done